Documentation/rust/quick-start.rst

```diff
@@ -3,97 +3,132 @@
 Quick Start
 ===========
 
-This document describes how to get started with Rust kernel development.
+This document describes how to get started with kernel development in Rust.
 
 
 Requirements
 ------------
 
-rustc and cargo
-***************
+This section explains how to fetch the requirements to work with Rust.
+If you have worked previously with Rust, this will only take a moment.
 
-A recent nightly Rust toolchain (with, at least, ``rustc`` and ``cargo``)
-is required, e.g. ``nightly-2021-01-02``. In the future, this restriction
-will be lifted.
+Some of these requirements might be available from your Linux distribution
+under names like ``rustc``, ``rust-src``, ``rust-bindgen``, etc. However,
+at the time of writing, they are likely to not be recent enough.
+
+
+rustc
+*****
+
+A recent *nightly* Rust toolchain (with, at least, ``rustc``) is required,
+e.g. ``nightly-2021-01-02``. Our goal is to use a stable toolchain as soon
+as possible, but for the moment we depend on a handful of nightly features.
 
 If you are using ``rustup``, run::
 
     rustup toolchain install nightly
 
-Otherwise, fetch a standalone installer from:
+Otherwise, fetch a standalone installer or install ``rustup`` from:
 
     https://www.rust-lang.org
 
 
-rustc sources
-*************
+Rust standard library source
+****************************
 
-The sources for the compiler are required to be available because the standard
-library (``core`` and ``alloc``) is cross-compiled.
+The Rust standard library source (``core`` and ``alloc``, at least) is required
+because the build system will cross-compile it.
 
 If you are using ``rustup``, run::
 
     rustup component add rust-src
 
-Otherwise, if you used a standalone installer, you can clone the Rust compiler
-sources and create a symlink to them in the installation folder of
-your nightly toolchain::
+Otherwise, if you used a standalone installer, you can clone the Rust
+repository into the installation folder of your nightly toolchain::
+
+    git clone https://github.com/rust-lang/rust `rustc --print sysroot`/lib/rustlib/src/rust
+
+
+compiler-builtins source
+************************
 
-    git clone https://github.com/rust-lang/rust
-    ln -s rust .../rust-nightly/lib/rustlib/src
+The source for ``compiler-builtins`` (a Rust port of LLVM's ``compiler-rt``)
+is required.
+
+The build system expects the sources alongside the Rust ones we just installed,
+so you can clone it into the installation folder of your nightly toolchain::
+
+    git clone https://github.com/rust-lang/compiler-builtins `rustc --print sysroot`/lib/rustlib/src/compiler-builtins
 
 
 bindgen
 *******
 
 The bindings to the C side of the kernel are generated at build time using
-``bindgen``. Currently we assume the latest version available, but that
-may change in the future.
+the ``bindgen`` tool. A recent version should work, e.g. ``0.57.0``.
```

Comment:
bindgen has to be compiled with a recent llvm version (nixOS still uses llvm 7 wich breakes). I think we should mention it

Reply:
Thanks for reminding me, will do.

Reply:
Hmm... the bindgen docs say libclang 3.9 or higher should work. Did bindgen break when building it with LLVM 7, or was it when invoking it to create the bindings? Anyway, it does not matter, since libclang needs to be quire recent to understand the kernel code.